### ISSUE TITLE
Watt-32 2.2 rev 10 locks up sometimes during tcp_tick() calls

### ISSUE THREAD

**Issue by mateuszviste** [gvanem/Watt-32#2]
I have created a real mode (16-bit) [DOS gopher client](http://gopherus.sourceforge.net) that uses Watt-32 for networking. It is built with OpenWatcom 1.9 and I am using Watt-32 2.2 rev 10, since the newer rev 11 isn't functioning at all on this prehistoric platform.

The application runs well most of the time, but with some very specific destinations it tends to lock up. After investigations, I found out that the freeze always happens when I call tcp_tick(socket). Further investigations lead me to discover that there is a chain of functions that ends with with a memcpy() call - and it's the memcpy() call that locks up the computer. Here's how it looks:

````
tcp_tick(s)
  _ip_handler()
    _tcp_handler()
      _tcp_fsm()
        tcp_estab_state()
          tcp_process_data()
            data_in_order()
              copy_in_order()
                memcpy()
````
The immediate reason is that memcpy() is performing a buffer overflow, because it is called with a value well outside the internal Watt-32 buffer. I attach a photo of my screen filled with the variety of printf calls that I added when investigating the problem. It shows, at the bottom of the screen, the arguments used by copy_in_order() when performing the memcpy() call.

![IMG_20200121_165116](https://user-images.githubusercontent.com/9503501/72825374-3408fb00-3c77-11ea-80a5-24daa2711695.jpg)

memcpy is being called with a negative length (-1024) which makes it actually copy 64512 bytes, since memcpy's argument is an unsigned int 16 bits wide.

in tcp_fsm.c I see there is such assumption in data_in_order():
````
  /* Skip data before recv_next. We must be left with some data or
   * we wouldn't have been called.
   */
  data += diff;
  len  -= diff;
````
This assumes that len > diff, while in my case it is not (len = 23, diff = 1047). I have no clue why is that nor I understand what "diff/ldiff" is representing exactly, nonetheless the code appears to be mishandling a condition it is not prepared to handle.

**Any pointers about what should I do next please?**

For the time being I added a primitive safe guard in tcp_process_data() that makes the function exit whenever len < ldiff ([PATCH HERE](https://github.com/gvanem/Watt-32/files/4094203/avoid_tcp_fsm_corruption_hack.zip)). This is unlikely to be a solution, it might perhaps corrupt the tcp stream even - I don't know, but at least it doesn't lock up my machine any more. A proper solution should be devised by someone who knows exactly the Watt-32 internals.

**Comment by gvanem** [comment]
It looks like you have the problem that Mikluas Patocka found a fix (?) for back in 2015
while debugging his [Web-browser](http://links.twibright.com/download/binaries/dos/).

Does your `tcp_fsm.c` have these lines that he contributed (from line 1075):
```c
  /* Check that *end* of packet is valid, i.e. will fit in advertised receive
   * window ('adv_win'). If it's before 'recv_next', we've seen it all before;
   * if it's after then the peer (or someone else) sent more than we said we
   * could take.
   */
#if 0
  if ((unsigned)len - ldiff > s->adv_win)
#else
  /*
   * Thanks to Mikulas Patocka <mikulas@twibright.com> for finding a problem
   * with the above 'if-test':
   *   if we receive a sequence of packets, all the packets are checked
   *   against 's->adv_win', but 's->adv_win' is not decreased as the packets are
   *   received, it remains the same. Consequently a packet that is out of window
   *   is errorneously accepted and corrupts memory.
   *
   * Hence this 'if-test' is used instead:
   */
  if ((unsigned)len - ldiff > s->max_rx_data - s->rx_datalen)
#endif
  {
    TCP_TRACE (("tcp_process_data (%u): packet ends outside %lu/%lu\n",
                __LINE__, (u_long)s->recv_next, (u_long)(s->recv_next + s->adv_win)));
    return (0);
  }
```

**Comment by mateuszviste** [comment]
Thank you for looking into this. I did see Mikluas' patch in the code yesterday, and I pondered about back-porting it to rel 10, but this does not seem to be it.

My understanding is that Mikluas solves another kind of problem - when a rogue destination sends data outside of Watt-32's TCP window. This is not what happens in my case - here WATT-32 advertises a window of 2K and the destination sends back a TCP segment with 23 bytes of payload. And somehow, this makes WATT-32 crash (because of the ldiff calculation that I do not grasp entirely, and that leads to a miscalculation of the destination buffer index).

I have performed a pcap capture this morning to be able to look at the wire level - unfortunately I'm still no wiser. [Here is the PCAP](https://github.com/gvanem/Watt-32/files/4096247/watt32_lockup1.zip). It contains a few connections to a specific gopher destination, I had to retry a few times before exhibiting the problem. The interesting packet that WATT-32 appears to trip on is the packet 179. I do not see what the problem is, but the destination definitely does respect the TCP window advertised by WATT-32.

**Comment by mateuszviste** [comment]
Looking at the PCAP I see nothing wrong with the communication at all. I also see no difference with the previous connections in the PCAP, that did not make WATT-32 crash. But I noticed that the problem seems to be linked to CPU speed (it appears almost every time on my 286 PC, but only 1 time out of few on my 386 PC). The frequency at which tcp_tick() is called seems to play a role here. The problem appears also much more often if I put lots of printf() debug stuff in tcp_tick().

We have two TCP segments incoming:
179:   23 bytes of data
180: 1024 bytes of data

both segments are transmitted close together (12 ms apart). From what I can understand by looking at my debug outputs, WATT-32 seems to confuse the two packets, because it says it got 23 bytes of TCP payload, but at an offset 1024 bytes in the future (ldiff = 1047, ie. 1024 + 23). I wonder whether it could be some kind of race condition, where I call tcp_tick() after WATT-32 obtained the 23-bytes segment, and during packet processing the pktdrvr feeds WATT-32 with the next 1024-bytes segment, leading to a mixup of both...? My theory is that when it works, it works because I manage to run tcp_tick() between the two packets. My 286 PC is slower, so it is more likely to be busy executing tcp_tick() while a new packet arrives while the previous one is still being processed.

Does that makes sense?

**Comment by gvanem** [comment]
> Does that make sense?

Sure. Since I've not tried using a 386 PC since the 90'ies, any guess could be correct. 
But I did not write this `ldiff` / TCP reassembly stuff.  So I do not fully understand it either.

I used `Wireshark watt32_lockup1.pcap` and looked at the **Analyze | Expert Information**. I saw there was a lot of re-transmissions. Could be related to the slow PC?

**Comment by mateuszviste** [comment]
> But I did not write this `ldiff` / TCP reassembly stuff. So I do not fully understand it either.

Ah, that's good to hear - I was starting to wonder whether I'm stupid or the code is indeed not so easy to grasp.

> I used `Wireshark watt32_lockup1.pcap` and looked at the **Analyze | Expert Information**. I saw there was a lot of re-transmissions. Could be related to the slow PC?

No, it is not. Even the 286 PC is a very fast machine when used correctly, and WATT-32 is certainly a very efficient beast.
There are two kinds of re-transmission in the PCAP. The first kind is WATT-32 becoming somewhat impatient re-transmitting the 2-bytes gopher query after roughly 80 ms of not getting an ACK (the remote server is far away, and the PC is connected over a laggy wifi link to a sluggish 128 Kbps land line). Nothing surprising there.

The second kind of re-transmission is what happens at the very end of the PCAP, starting at packet 181: the remote server notices that his answer is not being acquitted (because WATT-32 crashed in the mean time), so it re-sends its stuff several times before giving up.

Anyway, I will try to investigate the packet-driver-race-condition theory, perhaps I will stumble over something. And if not, then I will go back delving in tcp-fsm territories.

**Comment by mateuszviste** [comment]
Success! (I think)

The race-condition theory was wrong. I looked at the packet driver receiver, it has a one-packet buffer that is immediately handed over to queue mechanism while interrupts are disabled. No races there. The queue mechanism uses a 5-packets (on real mode targets) ring buffer, and from what I can grok, the author did take care (in 1998) to avoid changing data that is being accessed.

Now, I replaced all my debug printf()s by a single log that displays basic pkt-related data at _tcp_fsm() entry, and triggered the problem. Here it is:

![watt32_lockup3](https://user-images.githubusercontent.com/9503501/72911372-3c763a00-3d3a-11ea-837e-271056b9d690.jpg)

And [here is the PCAP](https://github.com/gvanem/Watt-32/files/4098307/watt32_lockup3.zip).

This made me realize that my initial analysis was wrong - I have been mislead because my screen was so full of debug garbage that I was seeing only 1 packet (the last one). This new test is much more obvious:

The server sends an answer within 2 packets (7 and 8 in the pcap):
````
pkt 7 = 23 bytes of TCP payload
pkt 8 = 1024 bytes of TCP payload
````
Watt32 receives each of these packets, processes them, but does not ack them. The fact that Watt32 does not ack the packets is not necessarily a problem itself - the TCP window is far from being full, so I assume (?) that Watt-32 delays the acks to save some peanuts on upload bandwidth. After about 340 ms, the remote server becomes worried that something got lost in transit, and sends packet 7 again (23 bytes). WATT-32 receives it and computes a difference (ldiff) of 1047 bytes (the payload being "behind" the current TCP window by that much bytes).
And that's where things become ugly - WATT-32 tries to add this data to the buffer, running into the negative-index-leads-to-buffer-overflow business, while what it should have done is simply ignoring this data since it has all been received already.

Long story short, my previous patch was actually doing the right thing, even though I didn't fully understand why. [HERE is a new, cleaner patch](https://github.com/gvanem/Watt-32/files/4098440/watt32_ignore_alreadyrcved_data.zip) (to be applied to the 2.2 rev 10 branch). It includes also Mikluas' patch. While Mikluas' patch is handling illegal data that tries to reach above the TCP window, it fails to detect the (more common, I would think) situation of old-and-processed-but-not-acked data being sent again, thus landing below the TCP window. My patch handles this case.

The other thing that one could be curious about, is why WATT-32 delays its ACK for over 340 ms... Seems quite longish. But this is something I will let someone else to figure out.

**Comment by gvanem** [comment]
Okay, that is good news. So a patch against *master* is now:
```diff
--- a/src/tcp_fsm.c
+++ b/src/tcp_fsm.c
@@ -1077,12 +1077,10 @@ static int tcp_process_data (_tcp_Socket *s, const tcp_Header *tcp,
    * if it's after then the peer (or someone else) sent more than we said we
    * could take.
    */
-#if 0
-  if ((unsigned)len - ldiff > s->adv_win)
-#else
+
   /*
    * Thanks to Mikulas Patocka <mikulas@twibright.com> for finding a problem
-   * with the above 'if-test':
+   * with the previous 'if-test':
    *   if we receive a sequence of packets, all the packets are checked
    *   against 's->adv_win', but 's->adv_win' is not decreased as the packets are
    *   received, it remains the same. Consequently a packet that is out of window
@@ -1090,8 +1088,19 @@ static int tcp_process_data (_tcp_Socket *s, const tcp_Header *tcp,
    *
    * Hence this 'if-test' is used instead:
    */
-  if ((unsigned)len - ldiff > s->max_rx_data - s->rx_datalen)
-#endif
+
+  /* First patch by Mikulas Patocka:
+   * checks that end of packet does not land above the TCP window
+   */
+  if ( ((unsigned)len - ldiff > s->max_rx_data - s->rx_datalen) ||
+      /*
+       * Second patch by Mateusz Viste:
+       * checks that end of packet does not land below the
+       * TCP window (legitimate case when remote peer retransmits an already-received
+       * segment that Watt-32 did not acquit due to delayed ack -
+       * Details at https://github.com/gvanem/Watt-32/issues/2
+       */
+      (ldiff > len) )
   {
     TCP_TRACE (("tcp_process_data (%u): packet ends outside %lu/%lu\n",
                 __LINE__, (u_long)s->recv_next, (u_long)(s->recv_next + s->adv_win)));
```

Not sure about the 340 msec delay. Have a try at the `tcp.timer.rto_base` etc. settings in `Wattcp.cfg`.

**Comment by gvanem** [comment]
I'll apply the above patch with credit to you.
